### PR TITLE
Feature/Explorer 38 - Update project using new backend features

### PR DESCRIPTION
**Pull request by VladislavPopovSR**
https://app.zenhub.com/workspaces/iroha-v2-60ddb820813b9100181fc060/issues/gh/soramitsu/iroha2-block-explorer-web/38

BaseTable & table.ts sticky pagination functionality might look messy right now. For now, it works as expected. Later I will implement query params & refactor.

### CODE REVIEW

File: src/shared/api/dto.ts [hunk 1]
```diff
@@ -1,183 +1,228 @@
...
+import { z } from 'zod';
+import BigNumber from 'bignumber.js';
+import invariant from 'tiny-invariant';
+
+const paginationSchema = z.object({
+  page: z.number(),
+  per_page: z.number(),
+  // TODO: remove nullable when backend is ready
+  total_pages: z.number().nullable(),
+  total_items: z.number().nullable(),
```
Comment (0x009922) [inline src/shared/api/dto.ts:10]: Not nullable anymore

File: src/shared/api/dto.ts [hunk 2]
```diff
@@ -1,183 +1,228 @@
...
-  interface Paginated<T> {
-    pagination: {
-      page: number
-      page_size: number
-      total: number
-    }
-    data: T[]
+class AssetDefinitionId {
+  private assetName = '';
+  private assetDomain = '';
```
Comment (0x009922) [inline src/shared/api/dto.ts:42]:
```suggestion
  private name: string;
  private domain: DomainId;
```

File: src/shared/api/dto.ts [hunk 3]
```diff
@@ -1,183 +1,228 @@
...
+    return this.assetName + '#' + this.assetDomain;
   }
+}
 
-  type AssetValue =
-    | Tagged<'Quantity', string>
-    | Tagged<'BigQuantity', string>
-    | Tagged<'Fixed', string>
-    | Tagged<'Store', JsonObject>;
+export function transformToAssetDefinitionId(assetDefinitionId: string): AssetDefinitionId {
```
Comment (0x009922) [inline src/shared/api/dto.ts:60]: In Zod, it would be idiomatic to use [`superRefine`](https://zod.dev/?id=superrefine), actually.

File: src/shared/api/dto.ts [hunk 4]
```diff
@@ -1,183 +1,228 @@
...
-  }
+  return new AssetDefinitionId(asset, asset_domain);
+}
 
-  type Mintable = 'Once' | 'Infinitely' | 'Not';
+class AssetId {
+  private assetName = '';
+  private assetDomain = '';
+  private accountSignatory = '';
+  private accountDomain = '';
```
Comment (0x009922) [inline src/shared/api/dto.ts:73]:
```suggestion
  private account: AccountId;
  private definition: AssetDefinitionId;
```

where both are also classes.

File: src/shared/api/dto.ts [hunk 5]
```diff
@@ -1,183 +1,228 @@
...
-    id: string
-    accounts: Account[]
-    asset_definitions: AssetDefinition[]
-    logo: null | string
-    metadata: any
-    /**
-     * amount of triggers, always 0 for now
-     */
-    triggers: number
+  public get account_domain() {
```
Comment (0x009922) [inline src/shared/api/dto.ts:91]: We don't have ESLint being setup properly, I guess, but it is common across the codebase to use `camelCase`.

```suggestion
  public get accountDomain() {
```

File: src/shared/ui/components/BaseTable.vue [hunk 6]
```diff
@@ -86,15 +86,16 @@ import { computed } from 'vue';
...
+import type { Pagination } from '@/shared/api/dto';
 
 interface Props {
   loading: boolean
-  pagination?: TablePagination | null
+  pagination?: Pagination | null
   items: T[]
   containerClass: string
   breakpoint?: string | number
+  sticky?: boolean
```
Comment (0x009922) [inline src/shared/ui/components/BaseTable.vue:98]: What _sticky_ means? Can it be a name that describes the behaviour better?

File: src/shared/ui/components/BaseTable.vue [hunk 7]
```diff
@@ -125,9 +126,23 @@ const segmentInfo = computed(() => {
...
+
+    const end = start - props.items.length + 1;
+
+    return `${start}—${end} of ${p.total_items}`;
+  }
+
+  const start = (p.page - 1) * p.per_page + 1;
+  const end = p.page * p.per_page;
+
+  return `${start}—${end > p.total_items ? p.total_items : end} of ${p.total_items}`;
```
Comment (0x009922) [inline src/shared/ui/components/BaseTable.vue:145]: This functionality seems a bit complicated. It would help to separate the complicated logic into some external function and cover it with unit tests thoroughly.

File: src/shared/ui/components/BaseTable.vue
Comment (0x009922) [inline src/shared/ui/components/BaseTable.vue:1]: This file in general has some non-trivial logic of direct/reverse pagination which is essential to test, imo. It could additionally help to make the implementation clearer.

But it's good to do in a separate PR, having a smaller scope.

File: src/shared/ui/composables/useErrorHandlers.ts [hunk 8]
```diff
@@ -18,5 +19,11 @@ export const useErrorHandlers = () => {
     notifications.error(errorMessage);
   }
 
-  return { handleUnknownError };
+  function handleZodError(error: ZodError) {
+    console.error(error);
+
+    notifications.error(error.errors[0].code);
+  }
```
Comment (0x009922) [inline src/shared/ui/composables/useErrorHandlers.ts:26]: It can be a single universal error handler that dynamically checks whether `error: unknown` is e.g. an `instanceof ZodError`.

Comment (0x009922) [reply src/shared/ui/composables/useErrorHandlers.ts:26]: Also, there is `.format` method on `ZodError`, please check the docs.

File: src/shared/api/dto.ts [hunk 9]
```diff
@@ -1,183 +1,228 @@
-/* eslint-disable no-unused-vars */
-/* eslint-disable camelcase */
-import type { JsonObject } from 'type-fest';
+import { z } from 'zod';
+import BigNumber from 'bignumber.js';
+import invariant from 'tiny-invariant';
+
+const paginationSchema = z.object({
```
Comment (0x009922) [inline src/shared/api/dto.ts:5]: Tip: in TypeScript, there are _runtime_ and _type_ independent namespaces. So, you can have a runtime item (variable, function) and a type item to have the same name.

Therefore, you can have a `const Pagination` that is a Zod schema, and a `type Pagination` that is `z.infer<typeof Pagination>`.

I would strongly advocate to this rename because having `xxxSchema` for all schemas is kind of verbose.

```suggestion
const Pagination = z.object({
```

File: src/shared/api/dto.ts
Comment (0x009922) [inline src/shared/api/dto.ts:1]: Would  make more sense to rename this module to `schemas.ts`.

File: src/shared/api/index.ts [hunk 10]
```diff
@@ -1,19 +1,30 @@
...
+} from '@/shared/api/dto';
+
 interface HTTPService {
   fetchAccounts: (params?: PaginationParams) => Promise<Paginated<Account>>
   fetchAccount: (id: string) => Promise<Account>
-  fetchAssets: (params?: PaginationParams) => Promise<Paginated<Asset>>
-  fetchAsset: (definition_id: string, account_id: string) => Promise<Asset>
-  fetchAssetDefinitions: (params: PaginationParams) => Promise<Paginated<AssetDefinition>>
+  fetchAssets: (params?: AssetSearchParams) => Promise<Paginated<Asset>>
+  fetchAsset: (id: string) => Promise<Asset>
```
Comment (0x009922) [inline src/shared/api/index.ts:19]: If you have a chance to use stronger types, please do!

```suggestion
  fetchAsset: (id: AssetId) => Promise<Asset>
```

File: src/shared/api/index.ts [hunk 11]
```diff
@@ -1,19 +1,30 @@
...
+  Paginated,
+  PaginationParams,
+  TransactionSearchParams,
+  Transaction,
+  PeerStatus,
+} from '@/shared/api/dto';
+
 interface HTTPService {
   fetchAccounts: (params?: PaginationParams) => Promise<Paginated<Account>>
   fetchAccount: (id: string) => Promise<Account>
```
Comment (0x009922) [inline src/shared/api/index.ts:17]:
```suggestion
  fetchAccount: (id: AccountId) => Promise<Account>
```

File: src/shared/api/index.ts [hunk 12]
```diff
@@ -1,19 +1,30 @@
...
-  fetchAssets: (params?: PaginationParams) => Promise<Paginated<Asset>>
-  fetchAsset: (definition_id: string, account_id: string) => Promise<Asset>
-  fetchAssetDefinitions: (params: PaginationParams) => Promise<Paginated<AssetDefinition>>
+  fetchAssets: (params?: AssetSearchParams) => Promise<Paginated<Asset>>
+  fetchAsset: (id: string) => Promise<Asset>
+  fetchAssetDefinitions: (params?: PaginationParams) => Promise<Paginated<AssetDefinition>>
   fetchAssetDefinition: (id: string) => Promise<AssetDefinition>
```
Comment (0x009922) [inline src/shared/api/index.ts:21]:
```suggestion
  fetchAssetDefinition: (id: AssetDefinitionId) => Promise<AssetDefinition>
```

File: src/shared/api/index.ts [hunk 13]
```diff
@@ -1,19 +1,30 @@
...
+  Block,
+  Domain,
+  Paginated,
+  PaginationParams,
+  TransactionSearchParams,
+  Transaction,
+  PeerStatus,
+} from '@/shared/api/dto';
+
 interface HTTPService {
```
Comment (0x009922) [inline src/shared/api/index.ts:15]: Actually, this whole interface doesn't bring benefit anymore. It was needed to ensure consistency between an actual API and a fake one.

So, it would make sense to re-export the `http.ts` module directly.

**Review by 0x009922 — CHANGES_REQUESTED** [review]

File: src/shared/ui/composables/useErrorHandlers.ts (hunk 8 above)
Comment (VladislavPopovSR) [reply src/shared/ui/composables/useErrorHandlers.ts:26]: I use `.parse` method instead of `.safeParse`. The following gives me an empty array:
```
error.format()._errors
```
I would prefer keep using `.parse` method so lets keep it as is. With that, we receive quite clear message about an error.

File: src/shared/api/dto.ts (hunk 9 above)
Comment (VladislavPopovSR) [reply src/shared/api/dto.ts:5]: Well, with such naming we can easy understand either it is a type or a zod object. Additionally, when I have
```
export type Account 
export const Account
```
and write `Account` typescript doesn't know what I want to import - type or schema. Hence, instead of auto import I have to check it's correctness every time.

The following might be useful:
```
export type Account
export const account
```

But I would keep it as is.

File: src/shared/ui/components/BaseTable.vue
Comment (VladislavPopovSR) [reply src/shared/ui/components/BaseTable.vue:1]: Agreed. Later I will separate it into table and pagination components. Also it will be easier to write unit tests.

File: src/shared/api/schemas.ts [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+
+export class DomainId {
+  private domain: string;
+  constructor(domain: string) {
+    this.domain = domain;
+  }
+  public getDomain() {
+    return this.domain;
+  }
+}
```
Comment (0x009922) [inline src/shared/api/schemas.ts:43]: `DomainId` doesn't have any logic and could be just a (branded) `string`.

File: src/shared/api/schemas.ts [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+
+export class AccountId {
+  private signatory: string;
+  private accountDomain: string;
```
Comment (0x009922) [inline src/shared/api/schemas.ts:47]:
```suggestion
  private domain: DomainId;
```

File: src/shared/api/schemas.ts [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+  }
+  public get assetDefinition() {
+    return this.definition.toString();
+  }
+  public get signatory() {
+    return this.account.getSignatory();
+  }
+  public get accountDomain() {
+    return this.account.getAccountDomain();
+  }
```
Comment (0x009922) [inline src/shared/api/schemas.ts:119]: Seems to be too verbose for no good reason.

It's fine to make the `account` and `definition` fields `readonly public` and access them directly.

Same applies to other classes.

File: src/shared/api/schemas.ts [hunk 17]
```diff
@@ -0,0 +1,308 @@
...
+  public toString() {
+    return this.definition.toString() + '#' + this.account.toString();
```
Comment (0x009922) [inline src/shared/api/schemas.ts:121]: The logic is a bit different.

If the definition's domain equals to the account's domain, the ID is shortened. Please refer to the API docs where I gave an example for `AssetId` model.

File: src/shared/api/schemas.ts [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+  const [asset, asset_domain] = assetDefinitionId.split('#');
+
+  return new AssetDefinitionId(asset, asset_domain);
+}
+
+export class AssetId {
+  private account: AccountId;
+  private definition: AssetDefinitionId;
+
+  public constructor(data: { asset: string, asset_domain: string, signatory: string, account_domain: string }) {
```
Comment (0x009922) [inline src/shared/api/schemas.ts:100]:
```suggestion
  public constructor(account: AccountId, definition: AssetDefinitionId) {
```

File: src/shared/api/dto.ts (hunk 9 above)
Comment (0x009922) [reply src/shared/api/dto.ts]: > and write `Account` typescript doesn't know what I want to import - type or schema

TypeScript perfectly knows what you are referring to depending on _where_ you are referring to it - in runtime or in type level.

> Hence, instead of auto import I have to check it's correctness every time.

What's the problem? It just a single `import { Account }` for both cases.

I would insist on this uniform naming.

However, a problem could occur is with classes such as `AccountId`, because classes are both types of the instance and also a value of the constructor. In this case having a `Schema` suffix is fine, imo.

File: src/shared/api/schemas.ts [hunk 19]
```diff
@@ -0,0 +1,308 @@
...
+    account_domain,
+  };
+
+  return new AssetId(data);
+}
+
+export const accountSchema = z.object({
+  id: z
+    .string()
+    .superRefine((val, ctx) => {
```
Comment (0x009922) [inline src/shared/api/schemas.ts:142]: Sorry for introducing confusion with `superRefine`. You can use `(val, ctx)` in `.transform` too, thus combining validation with parsing.

File: src/shared/api/schemas.ts [hunk 20]
```diff
@@ -0,0 +1,308 @@
...
+  hash: z.string(),
+  height: z.number(),
+  created_at: z.string().transform((x) => new Date(x)),
+  prev_block_hash: z.string().nullable(),
+  transactions_hash: z.string(),
+  transactions_rejected: z.number(),
+  transactions_total: z.number(),
+  consensus_estimation: z.object({
+    ms: z.number().min(0),
+  }),
```
Comment (0x009922) [inline src/shared/api/schemas.ts:291]: I would suggest making a separate `Duration` schema, as it might change in future and is used in several places.

```suggestion
  consensus_estimation: Duration,
```

Same for `Date`.

**Comment by 0x009922** [comment]
I have run frontend locally with the backend and collected some notes. Applicable both to frontend and backend. Leaving it here as a paper trail:

- Shift to `height`, not `hash`, as a primary Block ID, and use it everywhere. Return block hash only for `GET /blocks*` endpoints
- Account ID: shorten everywhere, **except** account details page (now the opposite)
- On each details page, print metadata as a raw JSON
- Blocks/Transactions/Instructions - handle "Age" uniformly. Print it as "12 secs ago" with a specific timestamp on hover (fine to do with HTML `title` attr for now)
- Transactions table: print authority and the kind of `instructions` (let's rename it to "Kind of Executable" for clarity) - is it `Instructions` or `WASM`
- Uniform table for Transactions and Instructions, everywhere:
	- `GET /transactions`: `authority`, `block` (height), `status`
	- `GET /instructions`: `authority`, `block`, `kind`
	- Has two sections in filters: Transactions (all, committed, rejected) and Instructions (all, then by kind)
- By pages
	- Blocks list: "Transactions" - add how many rejected ones, e.g. `4 (1 rejected)`
	- Account details
		- add table with owned domains
	- Assets page
		- currently fetches _assets_, not asset _definitions_; fix and print all relevant information from that endpoint
- Missing:
	- Transaction details page
	- Domain details page
	- TX filters aren't working yet

File: src/shared/api/schemas.ts [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+      definition.getDomain() ? definition.getDomain() : account.getAccountDomain()
+    );
+  }
+
+  public getAccount() {
+    return this.account;
+  }
+  public getDefintion() {
+    return this.definition;
+  }
```
Comment (0x009922) [inline src/shared/api/schemas.ts:111]: Please remove getters, make fields public, and access directly

File: src/shared/api/schemas.ts [hunk 22]
```diff
@@ -0,0 +1,299 @@
...
+    if (val.split('@').length !== 2) {
+      ctx.addIssue({
+        code: z.ZodIssueCode.invalid_literal,
+        message: 'Invalid accountId format',
+        expected: 'signatory@account_domain format',
+        received: val,
+      });
+    }
+
+    return transformToAccountId(val);
```
Comment (0x009922) [inline src/shared/api/schemas.ts:142]: No need for a dedicated `transformToAccountId` function, just introduces unnecessary jump in code. It could be inlined here.

File: src/shared/api/schemas.ts
Comment (0x009922) [inline src/shared/api/schemas.ts:1]: Naming between types and schemas is still not the same. Please make it so. (e.g. `pagination` vs `Pagination` to `Pagination` for both)

**Review by 0x009922 — CHANGES_REQUESTED** [review]

File: src/shared/api/schemas.ts (hunk 22 above)
Comment (VladislavPopovSR) [reply src/shared/api/schemas.ts:142]: I also use every such transform function in components. In example, `transformToAccountId` used in `AccountDetailsPage` in order to map route param. So it's better to have a reusable function.

File: src/shared/api/schemas.ts (hunk 16 above)
Comment (VladislavPopovSR) [reply src/shared/api/schemas.ts]: Done✅

File: src/shared/api/schemas.ts (hunk 15 above)
Comment (VladislavPopovSR) [reply src/shared/api/schemas.ts]: This is already resolved✅

File: src/shared/ui/components/BaseTable.vue (hunk 6 above)
Comment (VladislavPopovSR) [reply src/shared/ui/components/BaseTable.vue]: I have already renamed it to reversed✅

File: src/shared/api/schemas.ts (hunk 22 above)
Comment (0x009922) [reply src/shared/api/schemas.ts:121]: Reuse the schema itself!

File: src/pages/AccountDetailsPage.vue [hunk 23]
```diff
@@ -1,20 +1,20 @@
...
 import BaseHash from '@/shared/ui/components/BaseHash.vue';
 import TransactionStatus from '@/entities/transaction/TransactionStatus.vue';
-import { transactionModel } from '@/entities/transaction';
 import { useWindowSize } from '@vueuse/core';
 import { format } from '@/shared/lib/time';
 import BaseLoading from '@/shared/ui/components/BaseLoading.vue';
 import { useErrorHandlers } from '@/shared/ui/composables/useErrorHandlers';
-import invariant from 'tiny-invariant';
+import type { Account } from '@/shared/api/schemas';
+import { AccountIdSchema } from '@/shared/api/schemas';
```
Comment (0x009922) [inline src/pages/AccountDetailsPage.vue:17]:
```suggestion
import { type Account, AccountIdSchema } from '@/shared/api/schemas';
```

File: src/shared/api/schemas.ts [hunk 24]
```diff
@@ -0,0 +1,252 @@
...
+  id: AccountIdSchema,
+  metadata: Metadata,
+  owned_assets: z.number(),
+  owned_domains: z.number(),
+});
+
+export type Account = z.infer<typeof Account>;
+
+export interface AssetSearchParams extends PaginationParams {
+  owned_by?: string
```
Comment (0x009922) [inline src/shared/api/schemas.ts:162]:
```suggestion
  owned_by?: AccountId
```

Stronger types!

Tip: if you also add `toJSON()` method to `AccountId` class with `toString()` inside, it will automatically stringify on serialisation without manual intervention.

File: src/shared/api/schemas.ts [hunk 25]
```diff
@@ -0,0 +1,252 @@
...
+  metadata: Metadata,
+  owned_by: AccountIdSchema,
+  accounts: z.number(),
+  assets: z.number(),
+});
+
+export type Domain = z.infer<typeof Domain>;
+
+export interface TransactionSearchParams extends PaginationParams {
+  authority?: string
```
Comment (0x009922) [inline src/shared/api/schemas.ts:199]:
```suggestion
  authority?: AccountId
```

File: src/pages/AccountDetailsPage.vue [hunk 26]
```diff
@@ -27,43 +27,48 @@ const hashType = computed(() => (width.value < HASH_BREAKPOINT ? 'short' : 'medi
...
 onMounted(async () => {
   try {
     isFetchingAccount.value = true;
     account.value = await http.fetchAccount(accountId.value);
 
     if (account.value) {
-      await Promise.all([assetsTable.fetch(), transactionsTable.fetch()]);
+      await Promise.all([
+        assetsTable.fetch({ owned_by: accountId.value.toString() }),
+        transactionsTable.fetch({ authority: accountId.value.toString() }),
```
Comment (0x009922) [inline src/pages/AccountDetailsPage.vue:47]:
```suggestion
        assetsTable.fetch({ owned_by: accountId.value }),
        transactionsTable.fetch({ authority: accountId.value }),
```

File: src/pages/AssetDetailsPage.vue [hunk 27]
```diff
@@ -24,29 +25,22 @@ const { width } = useWindowSize();
 
 const hashType = computed(() => (width.value < HASH_BREAKPOINT ? 'medium' : 'full'));
 
-const assetName = computed(() => {
+const assetDefinitionId = computed(() => {
   const name = router.currentRoute.value.params['id'];
+  const rest = router.currentRoute.value.hash;
```
Comment (0x009922) [inline src/pages/AssetDetailsPage.vue:30]: You should find a way to put `#` in the route in some escaped way (urlencode). It should not be a part of `route.hash`.

File: src/pages/AssetDetailsPage.vue [hunk 28]
```diff
@@ -24,29 +25,22 @@ const { width } = useWindowSize();
...
 
-  invariant(typeof name === 'string', 'Expected string');
+  const assetId = AssetIdSchema.parse(name + rest);
```
Comment (0x009922) [inline src/pages/AssetDetailsPage.vue:32]: Why parsing `:id` in the route as `AssetId`? It is supposed to be `AssetDefinitionId`, e.g. `roses#wonderland`.

File: src/pages/AssetDetailsPage.vue [hunk 29]
```diff
@@ -59,13 +53,13 @@ onMounted(async () => {
 const transactionStatus = ref<ftm.Status>(null);
 
 // FIXME: this loads ALL transactions, not only related to the asset
-const transactionsTable = useTable(transactionModel.fetchList);
+const transactionsTable = useTable(http.fetchTransactions);
```
Comment (0x009922) [inline src/pages/AssetDetailsPage.vue:56]: Let's remove this table from the asset page altogether. We cannot implement it in the nearest future.

**Review by 0x009922 — CHANGES_REQUESTED** [review]

File: src/shared/api/schemas.ts (hunk 24 above)
Comment (VladislavPopovSR) [reply src/shared/api/schemas.ts]: Thanks, thats useful. By the way, I haven't used until yet approach to pass an object in the url so it's automatically uses toString method and converts our classes into normal strings.

File: src/pages/AccountDetailsPage.vue (hunk 23 above)
Comment (VladislavPopovSR) [reply src/pages/AccountDetailsPage.vue:17]: This is ESLint approach to separate imported types and values. Should I really manually change it or maybe there is new eslint rule?

Comment (0x009922) [reply src/pages/AccountDetailsPage.vue:17]: If ESLint does it automatically, than it's fine.

Maybe if we change rules it can patch it automatically though.

File: src/pages/AssetsList.vue [hunk 30]
```diff
@@ -24,14 +24,14 @@
       <template #row="{ item }">
         <div class="assets-list-page__row">
           <BaseLink
-            :to="`/assets/${item.id}`"
+            :to="`/assets/${encodeURIComponent(item.id.definition.toString())}`"
```
Comment (0x009922) [inline src/pages/AssetsList.vue:27]: As it is a common use case both for `AssetId` and `AssetDefinitionId`, maybe worth adding a method like `toStringEscaped()` to both classes. (though only definition id needs it)

**Review by 0x009922 — APPROVED** [review]
